### PR TITLE
Make watcher settings reloadable

### PR DESCRIPTION
**Pull request by hub-cap**
This commit allows for rebuilding watcher secure secrets via the
reload_secure_settings API call. The commit also renames a method in the
Notification Service to make it a bit more readable.

### CODE REVIEW

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: what about a `List<NotificationService>` that you fill in `createComponents`? This way you could get rid of the `getReloadableServices()` method entirely and loop through that list in the `reload()` method.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: how about naming this `reload()` as well just like in `ReloadablePlugin` plugin?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java]: is this PagerDutyService missing  

```
clusterSettings.addSettingsUpdateConsumer(this::loadSettings, getSettings());
```

in its ctor?

Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java]: In order to prevent this mistake, you could move

```
clusterSettings.addSettingsUpdateConsumer(this::loadSettings, getSettings());
```

into the `NotificationService` constructor and pass the `getSettings()` as arg to `super()`

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: watcher is not enabled, but reloading works? That could be a noop in this case, or?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: how about `isAccountCreated`?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: alternatively you could mock a NotificationService and check with `verify` if the `loadSettings` method was called, no need to have an impl then

**Review by spinscale — CHANGES_REQUESTED** [review]
left a few comments

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: The only reason i did not name it reload is because it is used to init settings as well, but ill change it to `reload`.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java]: I could also just have `NotificationService` impl a `getSettings` that the services overwrite. which is best?

Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java]: ahh, nm, its static, it would be more work to change all that.

Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java]: Im doing this in another commit, to ensure it does not muddy up this pull request. Ill edit this comment w/ the new PR once ive finished testing.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: in terms of the enable, there is no check to see if the service is active, like there are in things like register components. But there should be, since the case involving a reload for a inactive plugin will attempt to register some things and likely fail miserably. Will add & add tests, and just mock them so i dont have to impl them.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: no longer needed with the mock.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: 7be7b49

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: b97f21b

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: 153a97b

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: oops forgot to remove it!

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/PagerDutyService.java]: #31762 has been merged

**Comment by hub-cap** [comment]
These comments have been addressed, pls have another review @spinscale

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: `reloadableServices.forEach(NotificationService::reload)`?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: You can get rid of all the `SetOnce` instance definitions here again with the `List<>`, or?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: there is a `verifyNoInteractions()` method, that makes this test a bit easier to read

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: instead of exposing the `reloadableServices` via package private visibility, how about making it `protected` and do this (otherwise we have another different visibility in the watcher class)

```
        Watcher watcher = new Watcher(settings) {
            @Override
            public Collection<Object> createComponents(Client client, ClusterService clusterService, ThreadPool threadPool,
                                                       ResourceWatcherService resourceWatcherService, ScriptService scriptService,
                                                       NamedXContentRegistry xContentRegistry, Environment environment,
                                                       NodeEnvironment nodeEnvironment, NamedWriteableRegistry namedWriteableRegistry) {
                reloadableServices.add(mockService);
                return Collections.emptyList();
            }
        };

```

**Review by spinscale — CHANGES_REQUESTED** [review]
some more cosmetic changes, almost there

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: unneeded change

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: unneeded change

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: this is already documented in `ReloadablePlugin` I suppose?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: 👍

**Review by spinscale — APPROVED** [review]
left two minor cleanup comments. LGTM once CI passes

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java
Comment (spinscale) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/WatcherPluginTests.java]: 👍

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java
Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: Yea intellij automagically placed it here... i assumed that was a good thing (tm) but ill gladly remove it.

**Comment by hub-cap** [comment]
FUTURE ME, change the PR title to `watcher:....`

**Pull request merged**
